TORCH_FN

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39611 Unify TensorOptions signatures
* #39492 Move TensorOptions ops to c10
* #38361 Assert that kernels are called with the right signature
* **#39823 TORCH_FN**

Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

smessmer committed:
TORCH_FN

Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

ezyang commented:
This diff... is weirdly long for what I expected to see?

ezyang commented on aten/src/ATen/core/op_registration/op_registration.h:
```diff
@@ -240,6 +255,34 @@ class CAFFE2_API RegisterOperators final {
+     */
+    template<class FuncPtr>
+    // enable_if: only enable it if FuncType is actually a function
+    std::enable_if_t<c10::is_compile_time_function_pointer<FuncPtr>::value, Options&&> kernel(DispatchKey dispatch_key, FuncPtr) && {
```
Why are you adding more API surface to RegisterOperators? We should add `TORCH_FN` support to `torch/library.h` only and just port any sites which we want to use the new functionality.

smessmer replied:
I didn't want to port all call sites to the new API, that would have been an even larger PR. I need TORCH_FN for those cases though, so I added it to the old API too. Half of the functions I added here are scheduled to be removed very soon though (the unboxedOnly ones), so it's not that much of an API surface increase.

bhosmer replied:
Hmm, I'd echo not wanting to extend the old API. Unless TORCH_FN is urgent for some reason (is it? I actually thought it was an H2 thing) I think a natural dependency ordering is to port everything to the new API before doing anything that would require changes to the old.

I agree that the porting belongs in a different PR though. I mean, we should really do that anyway, regardless of new functionality. How much work is porting them, should we divvy them up to get them done?

smessmer replied:
I need TORCH_FN for the unboxing metaprogramming, that's why I did it now instead of waiting until we ported the API.

ezyang commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -1817,56 +1817,37 @@ static auto registry =
             torch::RegisterOperators::options()
-                .kernel<decltype(quantized_lstm_input), quantized_lstm_input>(
-                    DispatchKey::CPUTensorId))
+                .kernel(DispatchKey::CPUTensorId, TORCH_FN(quantized_lstm_input)))
```
You had to add a *lot* of code to handle this case, when you could have just ported.

smessmer replied:
For this specific case, I had to add one function with <10 lines. There's other call sites too and to support all of them, I had to add a bit more code, that's true. But porting all of them would be larger and I want to keep that a separate effort from this.

ezyang reviewed (changes requested):
We shouldn't be in the business of adding this function to RegisterOperators API. Notice that it was *way* more complicated to add support for this in RegisterOperators than in the new style API.

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

bhosmer commented:
> We shouldn't be in the business of adding this function to RegisterOperators API. 

+1, if there's compelling reasons to do this let's talk about them this afternoon. In general though, seems like upgrading the old API rather than decommissioning it would have to clear a pretty high bar.

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

smessmer commented:
ok I reduced this PR to only change the absolutely necessary call sites, this way I didn't have to add the feature to the old API. We'll have to add TORCH_FN to the other call sites in a separate PR now. cc @bhosmer @ezyang PTAL

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

ezyang commented on c10/core/CompileTimeFunctionPointer.h:
```diff
@@ -0,0 +1,49 @@
+
+}
+
+#define TORCH_FN_TYPE(func) ::c10::CompileTimeFunctionPointer<std::remove_pointer_t<std::remove_reference_t<decltype(func)>>, func>
```
For my knowledge: what's the benefit to removing reference and pointer types here?

smessmer replied:
Without doing this, only this works:

```cpp
constexpr auto fn = TORCH_FN(function)
```

With doing this, the above still works, but in addition also the following works:
```cpp
constexpr auto fn = TORCH_FN(&function)
```

which is a more natural syntax to get a function pointer.

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

smessmer committed:
Update on "TORCH_FN"


Add a compile time function pointer that can be used to pass function pointers in template args.
This is very useful for metaprogramming function wrappers.

Differential Revision: [D21986243](https://our.internmc.facebook.com/intern/diff/D21986243/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21986243/)!

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to have broken pytorch_windows_vs2017_14.11_py36_cpu_build and pytorch_windows_vs2017_14.11_py36_cuda10.1_build. Example log:

```
[1366/2816] C:\Users\circleci\project\build\win_tmp\bin\sccache-cl.exe   /TP -DAT_PARALLEL_OPENMP=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DUSE_EXTERNAL_MZCRC -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -Iinclude -Icaffe2\aten\src\TH -I..\aten\src\TH -Icaffe2\aten\src -I..\aten\..\third_party\catch\single_include -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\third_party\miniz-2.0.8 -I..\caffe2\core\nomnigraph\include -I..\caffe2\..\torch\csrc\api -I..\caffe2\..\torch\csrc\api\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include -I..\cmake\..\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\lib\site-packages\numpy\core\include -I..\cmake\..\third_party\pybind11\include -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -openmp -DNDEBUG -DUSE_VULKAN_WRAPPER -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG   -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -openmp -std:c++14 /showIncludes /Focaffe2\CMakeFiles\kernel_function_test.dir\__\aten\src\ATen\core\boxing\impl\kernel_function_test.cpp.obj /Fdcaffe2\CMakeFiles\kernel_function_test.dir\ /FS -c ..\aten\src\ATen\core\boxing\impl\kernel_function_test.cpp
FAILED: caffe2/CMakeFiles/kernel_function_test.dir/__/aten/src/ATen/core/boxing/impl/kernel_function_test.cpp.obj 
C:\Users\circleci\project\build\win_tmp\bin\sccache-cl.exe   /TP -DAT_PARALLEL_OPENMP=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DUSE_EXTERNAL_MZCRC -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -Iinclude -Icaffe2\aten\src\TH -I..\aten\src\TH -Icaffe2\aten\src -I..\aten\..\third_party\catch\single_include -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\third_party\miniz-2.0.8 -I..\caffe2\core\nomnigraph\include -I..\caffe2\..\torch\csrc\api -I..\caffe2\..\torch\csrc\api\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include -I..\cmake\..\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\lib\site-packages\numpy\core\include -I..\cmake\..\third_party\pybind11\include -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -openmp -DNDEBUG -DUSE_VULKAN_WRAPPER -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG   -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -openmp -std:c++14 /showIncludes /Focaffe2\CMakeFiles\kernel_function_test.dir\__\aten\src\ATen\core\boxing\impl\kernel_function_test.cpp.obj /Fdcaffe2\CMakeFiles\kernel_function_test.dir\ /FS -c ..\aten\src\ATen\core\boxing\impl\kernel_function_test.cpp
..\aten\src\ATen/core/op_registration/op_registration.h(238): error C2975: 'func_ptr_': invalid template argument for 'c10::CompileTimeFunctionPointer', expected compile-time constant expression
..\c10/core/CompileTimeFunctionPointer.h(32): note: see declaration of 'func_ptr_'
..\aten\src\ATen\core\boxing\impl\kernel_function_test.cpp(711): note: see reference to function template instantiation 'c10::RegisterOperators::Options &&c10::RegisterOperators::Options::kernel<Return(at::Tensor),Return `anonymous-namespace'::kernel_func<Return,at::Tensor>::func(at::Tensor)>(c10::DispatchKey) &&' being compiled
        with
        [
            Return=int64_t
        ]
..\c10/core/CompileTimeFunctionPointer.h(34): error C2338: TORCH_FN can only wrap function types.
..\aten\src\ATen/core/op_registration/op_registration.h(238): note: see reference to class template instantiation 'c10::CompileTimeFunctionPointer<Return (__cdecl *)(at::Tensor),0x0>' being compiled
        with
        [
            Return=int64_t
        ]
```
